### PR TITLE
added code for off chain publishing of metrics

### PR DESCRIPTION
**Pull request by PriyanshuDas**

### CODE REVIEW

File: src/contract.rs [hunk 1]
```diff
@@ -45,6 +45,11 @@ pub fn instantiate(
     };
     CONFIG.save(deps.storage, &config)?;
 
+    let off_chain_timestamps = OffChainTimestamps {
```
Comment (galacticminter) [inline src/contract.rs:48]: Some variant of this code need to be present in migrateMsg in a backwards compatible fashion. On a migrate message, we aren't going to run instantiate.

File: src/contract.rs [hunk 2]
```diff
@@ -101,8 +107,75 @@ pub fn execute(
...
+
+    Ok(Response::new()
+        .add_attribute("method", "remove_off_chain_metrics_for_timestamp")
+        .add_attribute("message",
+                       if timestamp_existed {"timestamp successfully removed"}
+                       else {"timestamp didn't exist"}))
+}
+
+//todo:test
+fn add_off_chain_metrics(deps: DepsMut, info: MessageInfo, metrics_to_be_added: OffChainMetrics)
```
Comment (galacticminter) [inline src/contract.rs:151]: 1. We will have upto a 130 validators in 2-3 weeks from now. I think we will hit the message too long error if we try take on all of this data in one message.

**Comment by galacticminter** [comment]
I imagined a simple (timestamp, validator) <=> (avg luna price, apr computed) mapping. 
Is this simpler than what you are trying to design?

**Comment by PriyanshuDas** [comment]
> 
I don't think I understand. We want the exchange ratio of different rewards currencies to luna to be stored right? What do you mean by mapping of a (timestmap, validator) with a luna price?

File: src/state.rs [hunk 3]
```diff
@@ -50,4 +59,33 @@ pub const STATE: Item<State> = Item::new("state");
...
+pub const OFF_CHAIN_VALIDATOR_IDX_MAPPING: Map<&Addr, u16> =
+    Map::new(constants::OFF_CHAIN_VALIDATOR_IDX_MAPPING);
+
+pub const OFF_CHAIN_TIMESTAMP_DETAILS: Map<U64Key, OffchainTimestampDetails> =
+    Map::new(constants::OFFCHAIN_TIMESTAMP_DETAILS);
+
+pub const OFF_CHAIN_STATE_FOR_VALIDATOR: Map<(U64Key, Addr), OffChainValidatorMetrics> =
+    Map::new(constants::OFF_CHAIN_METRICS_FOR_VALIDATOR);
+
+pub const OFF_CHAIN_TIMESTAMPS: Item<OffChainTimestamps> = Item::new("off_chain_timestamps");
```
Comment (galacticminter) [inline src/state.rs:76]: If I were you, I would change this to be a map as well. The key can be auto-incrementing (or maintained in the state). 
Screen Shot 2021-12-28 at 7 34 24 PM

Comment (galacticminter) [reply src/state.rs:76]: MAP IT ALLLLLL!

File: src/contract.rs [hunk 4]
```diff
@@ -101,13 +204,54 @@ pub fn execute(
...
+    -> Result<Response, ContractError> {
+
+    if !sender_is_manager(&deps, &info) {
+        return Err(ContractError::Unauthorized {});
+    }
+
+    let off_chain_timestamp_details =
+        OFF_CHAIN_TIMESTAMP_DETAILS.may_load(deps.storage, U64Key::from(timestamp));
+
+    println!("{:?}", off_chain_timestamp_details);
```
Comment (galacticminter) [inline src/contract.rs:238]: Clean this up

File: src/contract.rs [hunk 5]
```diff
@@ -101,13 +204,54 @@ pub fn execute(
...
+        }
+
+        //todo: test
+        ExecuteMsg::OffChainAddValidatorMetricsForTimestamp { timestamp, validator_metrics} => {
+            add_off_chain_validator_metrics(deps, info, timestamp, validator_metrics)
+        }
+    }
+}
+
+fn save_off_chain_details(
```
Comment (galacticminter) [inline src/contract.rs:227]: Can we rename this to save_offchain_timestamp_metadata ?

File: src/contract.rs [hunk 6]
```diff
@@ -914,13 +999,120 @@ fn query_validators_metrics_by_timestamp(
...
+        return Err(ContractError::Unauthorized {});
+    }
+
+    for validator_metric in metrics_to_be_added {
+        if off_chain_metrics_exists(&deps, timestamp, &validator_metric.opr_address) {
+            return Err(ContractError::OffChainMetricsAlreadyRecorded)
+        }
+        OFF_CHAIN_STATE_FOR_VALIDATOR.save(
+            deps.storage,
+            (U64Key::from(timestamp), validator_metric.opr_address.clone()),
```
Comment (galacticminter) [inline src/contract.rs:1053]: This should be (U64Key::from(timestamp), validator_index).

Get the validator index from the other mapping.

File: src/contract.rs [hunk 7]
```diff
@@ -914,13 +999,120 @@ fn query_validators_metrics_by_timestamp(
...
+
+    if !sender_is_manager(&deps, &info) {
+        return Err(ContractError::Unauthorized {});
+    }
+
+    let mut off_chain_timestamps = OFF_CHAIN_TIMESTAMPS.load(deps.storage)?;
+
+    let existing_timestamps_length = off_chain_timestamps.timestamps.len();
+
+    let new_timestamps = off_chain_timestamps.timestamps.into_iter()
```
Comment (galacticminter) [inline src/contract.rs:1017]: Simplify all of this logic by making timestamp a mapping between index <-> timestamp

Comment (galacticminter) [reply src/contract.rs:1017]: Overall the usage for remove_timestamp will be very limited.

File: src/contract.rs [hunk 8]
```diff
@@ -914,13 +999,120 @@ fn query_validators_metrics_by_timestamp(
...
+}
+
+//todo: implement
+fn get_off_chain_metrics(deps: Deps, timestamp: u64, validator_addr: Addr) -> StdResult<OffChainValidatorMetrics> {
+    let off_chain_state = OFF_CHAIN_STATE_FOR_VALIDATOR.load(
+        deps.storage, (U64Key::from(timestamp), validator_addr))?;
+    Ok(off_chain_state)
+}
+
+fn add_off_chain_validator(deps: DepsMut, info: MessageInfo, validator_addr: Addr)
```
Comment (galacticminter) [inline src/contract.rs:1080]: This needs to be part of the usual "add_validator" message. right?

Comment (galacticminter) [reply src/contract.rs:1080]: For now this can stay. But in the subsequent iteration, we should remove this message.

File: src/msg.rs [hunk 9]
```diff
@@ -90,3 +140,9 @@ pub struct ValidatorAprResponse {
     pub addr: Addr,
     pub apr: Decimal,
 }
+
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+#[serde(rename_all = "snake_case")]
+pub struct OffChainTimestamps {
```
Comment (galacticminter) [inline src/msg.rs:146]: Do you still need this?

File: src/state.rs [hunk 10]
```diff
@@ -13,7 +15,12 @@ pub struct State {
     pub validators: Vec<ValidatorAccounts>,
     //hard to remove from this, costs O(T) time, if was a set, could be O(1) average time
     pub cron_timestamps: Vec<u64>,
-    pub validator_index_for_next_cron: u64
+    pub validator_index_for_next_cron: u64,
+}
+
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+pub struct OffChainState {
+    pub new_validator_idx: u16,
```
Comment (galacticminter) [inline src/state.rs:23]: next_validator_idx?

File: src/msg.rs (hunk 9 above)
Comment (PriyanshuDas) [reply src/msg.rs:146]: Nope, will remove it.

File: src/state.rs (hunk 10 above)
Comment (PriyanshuDas) [reply src/state.rs:23]: Yup, will rename.

File: src/contract.rs [hunk 11]
```diff
@@ -914,9 +1014,163 @@ fn query_validators_metrics_by_timestamp(
...
+    if !OFF_CHAIN_TIMESTAMPS.has(deps.storage, U64Key::from(timestamp)) {
+        timestamp_existed = false;
+    }
+
+    OFF_CHAIN_TIMESTAMPS.remove(deps.storage, U64Key::from(timestamp));
+    OFF_CHAIN_TIMESTAMP_META_DATA.remove(deps.storage, U64Key::from(timestamp));
+
+    let validator_idxs_to_remove: Vec<u16> = OFF_CHAIN_STATE_FOR_VALIDATOR
+        .prefix(U64Key::from(timestamp))
+        .range(deps.storage, Option::None, Option::None, Order::Ascending)
```
Comment (galacticminter) [inline src/contract.rs:1040]: Can we paginate or delete say 30 validators at a time?
I believe this line will make the whole tx timeout.

File: src/contract.rs [hunk 12]
```diff
@@ -914,9 +1014,163 @@ fn query_validators_metrics_by_timestamp(
...
+        OFF_CHAIN_STATE_FOR_VALIDATOR
+            .remove(deps.storage, (U64Key::from(timestamp), U16Key::from(idx)))
+    });
+
+    Ok(Response::new()
+        .add_attribute("method", "remove_off_chain_metrics_for_timestamp")
+        .add_attribute(
+            "message",
+            if timestamp_existed {
+                "timestamp successfully removed"
```
Comment (galacticminter) [inline src/contract.rs:1054]: Understanding when this tx will be done is also useful especially as we paginate.

Comment (PriyanshuDas) [reply src/contract.rs:1054]: Yeah, will provide functionality to paginate this (if required). Currently just providing this as a msg if we want to clean up certain timestamps. Let's see how costly it is actually.

File: src/conversion_utils.rs [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+    let addr_string = vector.into_iter()
+        .map(|value| {value as char})
+        .collect::<String>();
+
+    Addr::unchecked(addr_string)
+}
+
+#[cfg(test)]
+mod tests {
+    // use crate::conversion_utils::{addr_from_vec_u8, u64_from_vec_u8};
```
Comment (galacticminter) [inline src/conversion_utils.rs:33]: Don't need these?

File: src/contract.rs [hunk 14]
```diff
@@ -36,15 +36,22 @@ pub fn instantiate(
...
     let config = Config {
         manager: info.sender.clone(),
         amount_to_stake_per_validator: msg.amount_to_stake_per_validator,
         batch_size: msg.batch_size,
     };
+
+    STATE.save(deps.storage, &state)?;
     CONFIG.save(deps.storage, &config)?;
 
+    // offchain publishing related states
```
Comment (galacticminter) [inline src/contract.rs:48]: Why don't you need this code?

File: src/contract.rs [hunk 15]
```diff
@@ -56,6 +63,85 @@ pub fn instantiate(
...
+    _deps: DepsMut,
+    _env: Env,
+    _msg: MigrateMsg,
+) -> Result<Response<TerraMsgWrapper>, ContractError> {
+    CONFIG.update(_deps.storage, |mut conf| -> StdResult<_> {
+        conf.manager = _msg.manager_address.clone();
+        Ok(conf)
+    })?;
+
+    // offchain publishing related states
```
Comment (galacticminter) [inline src/contract.rs:77]: Why don't you need this code?

File: src/contract.rs [hunk 16]
```diff
@@ -914,252 +970,679 @@ fn query_validators_metrics_by_timestamp(
...
 
-        let state = STATE.load(&dependencies.storage).unwrap();
-        let metrics = METRICS_HISTORY.load(
-            &dependencies.storage,
-            (
-                &Addr::unchecked(TEST_VALIDATOR_OPR_ADDR),
-                U64Key::from(TEST_TIMESTAMP_1),
-            ),
-        );
+    OFF_CHAIN_STATE.update(deps.storage, |mut s| -> StdResult<_> {
```
Comment (galacticminter) [inline src/contract.rs:1019]: You can do an OFF_CHAIN_STATE.save instead of update. Saves some gas.

File: src/contract.rs [hunk 17]
```diff
@@ -914,13 +970,229 @@ fn query_validators_metrics_by_timestamp(
...
+    for validator_metric in metrics_to_be_added {
+        let validator_idx = OFF_CHAIN_VALIDATOR_IDX_MAPPING
+            .load(deps.storage, &validator_metric.opr_address)
+            .unwrap();
+
+        if off_chain_metrics_exists(&deps, timestamp, validator_idx) {
+            return Err(ContractError::OffChainMetricsAlreadyRecorded);
+        }
+
+        OFF_CHAIN_STATE_FOR_VALIDATOR.save(
```
Comment (galacticminter) [inline src/contract.rs:1104]: You can combine the mayload call and save into a single `update` call.

**Review by galacticminter — APPROVED** [review]

**Pull request merged**